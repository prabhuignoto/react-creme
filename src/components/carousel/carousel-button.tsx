import classNames from "classnames";
import React from "react";
import { Button } from "..";
import { ChevronRightIcon } from "../../icons";
import "./carousel.scss";

interface CarouselButtonModel {
  onClick: () => void;
  position: "left" | "right";
}

const CarouselButton: React.FunctionComponent<CarouselButtonModel> = React.memo(
  ({ onClick, position }: CarouselButtonModel) => {
    return (
      <span
        className={classNames([
          "rc-carousel-btn",
          `rc-carousel-btn-${position}`,
        ])}
      >
        <Button type="icon" onClick={onClick}>
          <ChevronRightIcon />
        </Button>
      </span>
    );
  }
);

CarouselButton.displayName = "CarouselButton";

export { CarouselButton };
