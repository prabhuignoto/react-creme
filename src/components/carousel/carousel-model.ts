import React from "react";

export interface CarouselModel {
  children: React.ReactNode | React.ReactNode;
  direction?: "horizontal" | "vertical";
  height?: number;
}

export interface CarouselItemModel {
  height?: number;
  id: string;
  left?: string;
  top?: string;
  visible?: boolean;
  width?: number;
}

export interface CarouselItemsModel {
  activePage: number;
  carouselItems: CarouselItemModel[];
  children: React.ReactNode[] | React.ReactNode;
  direction: "horizontal" | "vertical";
  height: number;
  totalItems: number;
  width: number;
}
