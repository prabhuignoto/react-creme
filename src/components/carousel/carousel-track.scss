@use "../../design/layout.scss";
@use "../../design/list.scss";
@use "../../design/colors.scss";
@use "../../design/icon.scss";

.rc-carousel-track {
  &.rc-carousel-track-horizontal {
    @extend %center;
    @extend %list-horizontal;
  }

  &.rc-carousel-track-vertical {
    @extend %list;
    width: 3rem;
  }
}

.rc-carousel-track-item {
  @extend %shadow-inset;
  border: 1px solid colors.$secondary;
  border-radius: 50%;
  height: 10px;
  margin: 4px;
  width: 10px;
}

.rc-carousel-track-item:hover {
  cursor: pointer;
}

.rc-carousel-track-item-selected {
  background: colors.$primary;
  border: 1px solid colors.$primary;
}

.rc-carousel-icon {
  @extend %icon-sm;
}

.rc-carousel-track-wrapper-horizontal {
  @extend %center;
}

.rc-carousel-track-wrapper-vertical {
  @extend %list;
}
