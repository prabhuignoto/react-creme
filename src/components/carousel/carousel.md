| props     | description                                                       | datatype  | default      |
| :-------- | :---------------------------------------------------------------- | :-------- | :----------- |
| children  | content that will be embedded in the accordion body               | ReactNode |              |
| direction | Orientation of the carousel                                       | String    | "Horizontal" |
| height    | Minimum height of the carousel                                    | Number    | 400          |
| autoplay  | Autoplays the carousel on load. Specify the interval as the value | Number    | 0            |