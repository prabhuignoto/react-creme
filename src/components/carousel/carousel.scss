@use "../../design/button.scss";
@use "../../design/icon.scss";
@use "../../design/colors.scss";
@use "../../design/list.scss";
@use "../../design/layout.scss";

.carousel-container {
  @extend %border;
  @extend %shadow-inset;
  width: 100%;

  &.carousel-container-horizontal {
    @extend %col;
  }

  &.carousel-container-vertical {
    @extend %center;
  }
}

.carousel-wrapper {
  min-height: var(--min-height);
  overflow: hidden;
  position: relative;
  width: calc(100% - 3rem);
}

.carousel-track-wrapper {
  @extend %center;
  padding: 0.5rem 0;

  &.carousel-track-wrapper-horizontal {
    width: 100%;
  }
}

.carousel {
  height: 100%;
  list-style: none;
  margin: 0;
  padding: 0;
  position: absolute;
  transition: transform 0.3s cubic-bezier(0.215, 0.61, 0.355, 1);
}

.carousel-item {
  @extend %center;
  position: absolute;
}

.carousel-btn {
  @extend %button-lg-no-border;
  position: absolute;
  top: 50%;
  transform: translateY(-50%);
  z-index: 100;
}

.carousel-btn-left {
  left: 1rem;
  transform: rotate(180deg) translateY(50%);
}

.carousel-btn-right {
  right: 1rem;
}
