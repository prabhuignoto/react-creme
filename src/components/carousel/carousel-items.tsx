import React, { CSSProperties, ReactNode, useMemo } from "react";
import { CarouselItemsModel } from "./carousel-model";
import "./carousel.scss";

const CarouselItems: React.FunctionComponent<CarouselItemsModel> = ({
  activePage = 0,
  carouselItems,
  children,
  direction,
  height,
  totalItems,
  width,
}: CarouselItemsModel) => {
  const carouselStyle = useMemo(() => {
    return {
      width: `${width * totalItems}px`,
      transform:
        direction === "horizontal"
          ? `translateX(-${width * activePage}px)`
          : `translateY(-${height * activePage}px)`,
    } as CSSProperties;
  }, [width, height, activePage]);

  return (
    <ul className={"rc-carousel"} style={carouselStyle} role="list">
      {carouselItems.map((item, index) => (
        <li
          key={item.id}
          className={"rc-carousel-item"}
          role="listitem"
          style={{
            width: item.width,
            height: direction === "horizontal" ? "100%" : item.height - 32,
            top: direction === "horizontal" ? 0 : item.top,
            left: direction === "horizontal" ? item.left : 0,
          }}
        >
          <div className="rc-carousel-item-container">
            {(children as ReactNode[])[index]}
          </div>
        </li>
      ))}
    </ul>
  );
};

CarouselItems.displayName = "CarouselItems";

export { CarouselItems };
