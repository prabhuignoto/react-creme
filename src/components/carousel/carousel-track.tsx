import classNames from "classnames";
import React, { useMemo } from "react";
import { CarouselButton } from "./carousel-button";
import { CarouselTrackModel } from "./carousel-model";
import "./carousel-track.scss";

const CarouselTrack: React.FunctionComponent<CarouselTrackModel> = ({
  length,
  handleSelection,
  activeIndex,
  direction,
  onNext,
  onPrevious,
  hidePrevious,
  hideNext,
}: CarouselTrackModel) => {
  const carouselTrackClass = useMemo(
    () => classNames(["rc-carousel-track", `rc-carousel-track-${direction}`]),
    []
  );

  const carouselContainerClass = useMemo(
    () =>
      classNames("rc-carousel-track-wrapper", {
        [`rc-carousel-track-wrapper-${direction}`]: true,
      }),
    [direction]
  );

  return (
    <div className={carouselContainerClass}>
      <CarouselButton
        onClick={onPrevious}
        position="left"
        hide={hidePrevious}
        direction={direction}
      />
      <ul className={carouselTrackClass} role="list">
        {Array.from({ length }).map((_, index) => (
          <li
            key={index}
            role="listitem"
            className={classNames([
              "rc-carousel-track-item",
              index === activeIndex ? "rc-carousel-track-item-selected" : "",
            ])}
            onClick={() => handleSelection(index)}
          ></li>
        ))}
      </ul>
      <CarouselButton
        onClick={onNext}
        position="right"
        hide={hideNext}
        direction={direction}
      />
    </div>
  );
};

export { CarouselTrack };
