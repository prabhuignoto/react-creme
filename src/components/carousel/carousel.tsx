import classNames from "classnames";
import { nanoid } from "nanoid";
import React, {
  CSSProperties,
  startTransition,
  useCallback,
  useEffect,
  useLayoutEffect,
  useMemo,
  useRef,
  useState,
} from "react";
import { useDebounce } from "use-debounce";
import useSwipe from "../common/effects/useSwipe";
import { CarouselItems } from "./carousel-items";
import { CarouselItemModel, CarouselModel } from "./carousel-model";
import { CarouselTrack } from "./carousel-track";
import "./carousel.scss";
import ResizeObserver from "resize-observer-polyfill";

const Carousel: React.FunctionComponent<CarouselModel> = ({
  autoPlay = 0,
  children,
  direction = "horizontal",
  height = 400,
  transition = "cubic-bezier(0.55, 0.08, 0.68, 0.53)",
  enableSwipe = false,
}) => {
  const carouselRef = useRef<HTMLDivElement | null>(null);
  const [carouselItems, setCarouselItems] = useState<CarouselItemModel[]>(
    Array.isArray(children)
      ? children.map(() => ({
          id: nanoid(),
          visible: false,
          height: 0,
          width: 0,
        }))
      : []
  );

  const trackCount = useRef(Array.isArray(children) ? children.length : 0);
  const autoPlayRef = useRef<number>();

  const [activePage, setActivePage] = useState(0);
  const [slideWidth, setSlideWidth] = useState(0);
  const [slideHeight, setSlideHeight] = useState(0);
  const [isAutoPlaying, setAutoPlaying] = useState(!!autoPlay);

  const [debouncedSlideWidth] = useDebounce(slideWidth, 100);
  const [debouncedSlideHeight] = useDebounce(slideHeight, 100);

  const resizeObserver = useRef<ResizeObserver | null>(null);

  const handleNext = useCallback(() => {
    activePage < trackCount.current - 1 &&
      startTransition(() => setActivePage((prev) => prev + 1));
  }, [activePage]);

  const handlePrevious = useCallback(
    () =>
      activePage > 0 &&
      startTransition(() => setActivePage((prev) => prev - 1)),
    [activePage]
  );

  const handleActivatePage = useCallback(
    (pageIndex) => setActivePage(pageIndex),
    []
  );

  const onInitRef = useCallback((node) => {
    if (node) {
      const { clientHeight, clientWidth } = node;
      setSlideWidth(clientWidth);
      setSlideHeight(clientHeight);

      resizeObserver.current = new ResizeObserver(() => {
        const { clientHeight, clientWidth } = node as HTMLElement;
        startTransition(() => {
          setSlideWidth(clientWidth);
          setSlideHeight(clientHeight);
        });
      });

      resizeObserver.current.observe(node);
    }
  }, []);

  useLayoutEffect(() => {
    if (slideWidth && slideHeight && Array.isArray(children)) {
      const prop = direction === "horizontal" ? "left" : "top";
      setCarouselItems(
        children.map((_, index) => ({
          id: nanoid(),
          width: debouncedSlideWidth,
          height: debouncedSlideHeight,
          visible: false,
          [prop]: `${
            direction === "horizontal"
              ? index * slideWidth
              : index * slideHeight
          }px`,
        }))
      );
    }
  }, [debouncedSlideWidth, debouncedSlideHeight]);

  const hideNextButton = useMemo(
    () => activePage === trackCount.current - 1,
    [activePage]
  );

  useLayoutEffect(() => {
    if (
      !autoPlayRef.current &&
      autoPlay &&
      activePage < carouselItems.length - 1
    ) {
      setAutoPlaying(true);
      autoPlayRef.current = setInterval(() => {
        setActivePage((page) => {
          if (page === carouselItems.length - 1) {
            clearInterval(autoPlayRef.current);
            setAutoPlaying(false);
            return page;
          }

          return page + 1;
        });
      }, autoPlay);
    }
  }, [autoPlay, carouselItems.length]);

  const hidePreviousButton = useMemo(() => activePage === 0, [activePage]);

  const carouselTrackClass = useMemo(
    () =>
      classNames([
        "rc-carousel-track-wrapper",
        `rc-carousel-track-wrapper-${direction}`,
        {
          "rc-carousel-track-auto-play": isAutoPlaying,
        },
      ]),
    [isAutoPlaying]
  );

  const carouselContainerClass = useMemo(
    () =>
      classNames([
        "rc-carousel-container",
        `rc-carousel-container-${direction}`,
      ]),
    []
  );

  const carouselWrapperStyle = useMemo(
    () =>
      ({
        "--min-height": `${height}px`,
        "--transition": transition,
      } as CSSProperties),
    []
  );

  const wrapperClass = useMemo(
    () =>
      classNames(["rc-carousel-wrapper", `rc-carousel-wrapper-${direction}`]),
    []
  );

  if (enableSwipe) {
    const { dir, offset } = useSwipe(carouselRef, "low");

    useEffect(() => {
      if (
        (dir === "RIGHT" && direction === "horizontal") ||
        (dir === "BOTTOM" && direction === "vertical")
      ) {
        handlePrevious();
      } else if (
        (dir === "LEFT" && direction === "horizontal") ||
        (dir === "TOP" && direction === "vertical")
      ) {
        handleNext();
      }
    }, [offset]);
  }

  useEffect(() => {
    return () => {
      resizeObserver.current?.disconnect();
    };
  }, []);

  return (
    <div className={carouselContainerClass}>
      <div
        className={wrapperClass}
        ref={onInitRef}
        style={carouselWrapperStyle}
      >
        <CarouselItems
          carouselItems={carouselItems}
          direction={direction}
          activePage={activePage}
          totalItems={trackCount.current}
          height={slideHeight}
          width={slideWidth}
        >
          {children}
        </CarouselItems>
      </div>
      <div className={carouselTrackClass}>
        {!isAutoPlaying && (
          <CarouselTrack
            activeIndex={activePage}
            handleSelection={handleActivatePage}
            length={trackCount.current}
            direction={direction}
            onNext={handleNext}
            onPrevious={handlePrevious}
            hideNext={hideNextButton}
            hidePrevious={hidePreviousButton}
          />
        )}
      </div>
    </div>
  );
};

export { Carousel };
