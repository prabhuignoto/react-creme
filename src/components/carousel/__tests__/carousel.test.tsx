import { fireEvent, render, waitFor } from "@testing-library/react";
import React from "react";
import { act } from "react-dom/test-utils";
import { Carousel } from "../carousel";

describe("Carousel", () => {
  it("should render carousel", () => {
    const { getByText } = render(
      <Carousel direction="horizontal">
        <span>one</span>
        <span>two</span>
      </Carousel>
    );

    expect(getByText("one")).toBeInTheDocument();
    expect(getByText("two")).toBeInTheDocument();
  });

  it("should render vertical carousel", () => {
    const { getByText } = render(
      <Carousel direction="vertical">
        <span>one</span>
        <span>two</span>
      </Carousel>
    );

    expect(getByText("one")).toBeInTheDocument();
    expect(getByText("two")).toBeInTheDocument();
  });

  it("should carousel navigation work as expected", async () => {
    const { container } = render(
      <Carousel>
        <span>one</span>
        <span>two</span>
      </Carousel>
    );

    if (container) {
      expect(
        container.querySelector(".carousel-btn-right")
      ).toBeInTheDocument();

      const btnRight = container.querySelector(".carousel-btn-right");
      const btnLeft = container.querySelector(".carousel-btn-left");
      const carouselItems = container.querySelectorAll(".carousel-item");

      if (btnRight && carouselItems) {
        await act(async () => {
          fireEvent.click(btnRight);
        });

        await waitFor(async () => {
          expect(carouselItems[1]).toHaveStyle("visibility: visible");
        });
      }

      if (btnLeft && carouselItems) {
        await act(async () => {
          fireEvent.click(btnLeft);
        });

        await waitFor(async () => {
          expect(carouselItems[0]).toHaveStyle("visibility: visible");
        });
      }
    }
  });
});
