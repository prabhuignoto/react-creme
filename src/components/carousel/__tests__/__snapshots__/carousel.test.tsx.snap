// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should render carousel snapshot 1`] = `
<div
  class="rc-carousel-container rc-carousel-container-horizontal"
>
  <div
    class="rc-carousel-wrapper rc-carousel-wrapper-horizontal"
    style="--min-height: 400px;"
  >
    <ul
      class="rc-carousel"
      role="list"
      style="width: 0px; transform: translateX(-0px);"
    >
      <li
        class="rc-carousel-item"
        role="listitem"
        style="width: 0px; top: 0px;"
      >
        <div
          class="rc-carousel-item-container"
        >
          <span>
            one
          </span>
        </div>
      </li>
      <li
        class="rc-carousel-item"
        role="listitem"
        style="width: 0px; top: 0px;"
      >
        <div
          class="rc-carousel-item-container"
        >
          <span>
            two
          </span>
        </div>
      </li>
    </ul>
  </div>
  <div
    class="rc-carousel-track-wrapper rc-carousel-track-wrapper-horizontal"
  >
    <div
      class="rc-carousel-track-wrapper rc-carousel-track-wrapper-horizontal"
    >
      <span
        class="rc-carousel-btn rc-carousel-btn-left rc-carousel-btn-horizontal rc-carousel-btn-hide"
      >
        <button
          class="rc-btn-sm rc-btn-icon"
          role="button"
          tabindex="0"
        >
          <span
            class="rc-btn-icon-container"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 18l6-6-6-6"
              />
            </svg>
          </span>
        </button>
      </span>
      <ul
        class="rc-carousel-track rc-carousel-track-horizontal"
        role="list"
      >
        <li
          class="rc-carousel-track-item rc-carousel-track-item-selected"
          role="listitem"
        />
        <li
          class="rc-carousel-track-item"
          role="listitem"
        />
      </ul>
      <span
        class="rc-carousel-btn rc-carousel-btn-right rc-carousel-btn-horizontal"
      >
        <button
          class="rc-btn-sm rc-btn-icon"
          role="button"
          tabindex="0"
        >
          <span
            class="rc-btn-icon-container"
          >
            <svg
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M9 18l6-6-6-6"
              />
            </svg>
          </span>
        </button>
      </span>
    </div>
  </div>
</div>
`;
